agent/src/main/java/com/teamscale/jacoco/agent/testimpact/TestwiseCoverageAgent.java

```diff
@@ -71,21 +144,115 @@ private String handleTestStart(Request request, Response response) {
 	private String handleTestEnd(Request request, Response response) throws DumpException {
 		String testId = request.params(TEST_ID_PARAMETER);
 		if (testId == null || testId.isEmpty()) {
-			logger.error("Test name missing in " + request.url() + "!");
-
-			response.status(400);
-			return "Test name is missing!";
+			return error("Test name missing in " + request.url() + "!", response);
 		}
 
+		notifyTestEnd(testId);
 		logger.debug("End test " + testId);
+
 		controller.dump();
 
 		response.status(204);
 		return "";
 	}
 
+	/** Notifies all secondary agents about a test start event. */
+	private void notifyTestStart(String testId) {
+		secondaryAgents.forEach((port, agent) -> agent.signalTestStart(testId).enqueue(
+				IAgentService.failureCallback(() -> logger.warn("Error signaling test start to agent on port " + port))));
+	}
+
+	/** Notifies all secondary agents about a test end event. */
+	private void notifyTestEnd(String testId) {
+		secondaryAgents.forEach((port, agent) -> agent.signalTestEnd(testId).enqueue(
+				IAgentService.failureCallback(() -> logger.warn("Error signaling test end to agent on port " + port))));
+	}
+
+	/** Handles registrations from secondary agents. */
+	private String handleRegister(Request request, Response response) {
+		if (primaryAgent != null) {
+			return error("Cannot register with a secondary agent.", response);
+		}
+
+		try {
+			int port = extractPortFromQuery(request);
+			secondaryAgents.put(port, IAgentService.create(port));
+		} catch (IllegalArgumentException e) {
+			return error(e.getMessage(), response);
+		}
+
+		response.status(204);
+		return "";
+	}
+
+	/** Handles unregistrations from secondary agents. */
+	private String handleUnregister(Request request, Response response) {
+		if (primaryAgent != null) {
+			return error("Cannot unregister from a secondary agent.", response);
+		}
+
+		try {
+			int port = extractPortFromQuery(request);
+			IAgentService previous = secondaryAgents.remove(port);
+			if (previous == null) {
+				return error("No secondary agent registered for port " + port, response);
+			}
+		} catch (IllegalArgumentException e) {
+			return error(e.getMessage(), response);
+		}
+		response.status(204);
+		return "";
+	}
+
+	/**
+	 * Extracts and validates the port number from the request.
+	 * 
+	 * @throws IllegalArgumentException
+	 *             if the port number is missing, not a number, or an invalid
+	 *             port.
+	 */
+	private static int extractPortFromQuery(Request request) throws IllegalArgumentException {
+		String portString = request.queryParams("port");
+		if (portString == null || portString.isEmpty()) {
+			throw new IllegalArgumentException("Port is missing!");
+		}
+		int port = Integer.parseInt(portString);
+		if (port < 1024 || port >= MAX_PORT_NUMBER) {
+			throw new NumberFormatException("Port " + port + " is not a valid port.");
```

Comment:
Same here

Reply:
I thought since `Integer.parseInt` already throws `NumberFormatException`, which extends `IllegalArgumentException` I use this here as well and can catch all of them at once in the calling method, since they are all related.

Reply:
Makes sense